src/wallet/fidelity.rs

```diff
@@ -15,456 +38,582 @@ const BOND_VALUE_EXPONENT: f64 = 1.3;
+            .get(&index)
+            .ok_or(FidelityError::BondDoesNotExist)?;
+        let current_time = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("This can't error")
+            .as_secs();
+
+        let hash = self.rpc.get_block_hash(bond.conf_height as u64)?;
+
+        let confirmation_time = self.rpc.get_block_header_info(&hash)?.time as u64;
+
+        let locktime = match bond.lock_time {
+            LockTime::Blocks(blocks) => {
+                let tip_hash = self.rpc.get_blockchain_info()?.best_block_hash;
+                let (tip_height, tip_time) = {
+                    let info = self.rpc.get_block_header_info(&tip_hash)?;
+                    (info.height, info.time as u64)
+                };
+                // Estimated locktime from block height = [current-time + (maturity-height - block-count) * 10 * 60] sec
+                tip_time + ((blocks.to_consensus_u32() - tip_height as u32) * 10 * 60) as u64
+            }
+            LockTime::Seconds(sec) => sec.to_consensus_u32() as u64,
+        };
 
-        let onion_msg_hash =
-            Message::from_slice(sha256d::Hash::hash(onion_hostname.as_bytes()).as_byte_array())
-                .unwrap();
-        let onion_sig = secp.sign_ecdsa(&onion_msg_hash, &cert_privkey);
+        let bond_value = calculate_fidelity_value(
+            Amount::from_sat(bond.amount),
+            locktime,
+            confirmation_time,
+            current_time,
+        );
 
-        FidelityBondProof {
-            utxo: self.utxo,
-            utxo_key: self.utxo_key,
-            locktime: self.locktime,
-            cert_sig,
-            cert_expiry,
-            cert_pubkey,
-            onion_sig,
-        }
+        Ok(bond_value)
     }
-}
 
-impl FidelityBondProof {
-    /// Verifies the fidelity bond proof and gets the transaction output.
-    pub fn verify_and_get_txo(
-        &self,
-        rpc: &Client,
-        block_count: u64,
-        onion_hostname: &str,
-    ) -> GetTxOutResult {
-        let secp = Secp256k1::new();
+    /// Create a new fidelity bond with given amount and locktime.
+    /// This functions creates the fidelity transaction, signs and broadcast it.
+    /// Upon confirmation it stores the fidelity information in the wallet data.
+    pub fn create_fidelity(
+        &mut self,
+        amount: Amount,
+        locktime: LockTime, // The final locktime in blockheight or timestamp
+    ) -> Result<u32, WalletError> {
+        let (index, fidelity_addr, fidelity_pubkey) = self.get_next_fidelity_address(locktime)?;
+
+        // Fetch utxos, filter out existing fidelity coins
+        let mut unspents = self
+            .list_unspent_from_wallet(false, false)?
+            .into_iter()
+            .filter(|(_, spend_info)| !matches!(spend_info, UTXOSpendInfo::FidelityBondCoin { .. }))
+            .collect::<Vec<_>>();
+
+        unspents.sort_by(|a, b| b.0.amount.cmp(&a.0.amount));
```

Comment:
Any specific reason to sort in desc,? As both sorted and non-sorted unspents will result in sub optimal solutions. Do we want to have selected utxo size to be as low as possible?

Reply:
For the dumb coinselection I did, it's easier to write it if it's sorted.. Its suboptimal anyway.. We won't be targeting wallet optimization here anyway.. 

And yes the sort allow us to have lowest selection size. Not that its super important.. 

corollary idea: having an independent rust coinselection library would be super awesome..